From 7f4f06c942cb27af7f7be05459fd4d94d70c4135 Mon Sep 17 00:00:00 2001
From: Iurii Konovalenko <iurii.konovalenko@globallogic.com>
Date: Tue, 1 Dec 2015 20:22:17 +0200
Subject: [PATCH] libxl: Add comatible list to config file

Some platforms needs more compatible properety vaulues in device
tree root node in addition to "xen,xenvm-%d.%d" and "xen,xenvm"
values, that are given to  by default.
Specify in domain configuration file which values should be added,
writing in properety "compatibles" compat string, separated by comas.

Change-Id: I05c48cf0d568dd2de08894cc8cd18233e6fa5fb1
Signed-off-by: Iurii Konovalenko <iurii.konovalenko@globallogic.com>
---
 tools/libxl/libxl_arm.c     | 34 ++++++++++++++++++++++++++--------
 tools/libxl/libxl_types.idl |  5 +++++
 tools/libxl/xl_cmdimpl.c    | 14 +++++++++++++-
 3 files changed, 39 insertions(+), 9 deletions(-)

diff --git a/tools/libxl/libxl_arm.c b/tools/libxl/libxl_arm.c
index f00f7bb..e3bf135 100644
--- a/tools/libxl/libxl_arm.c
+++ b/tools/libxl/libxl_arm.c
@@ -231,20 +231,38 @@ static int fdt_property_regs(libxl__gc *gc, void *fdt,
 
 static int make_root_properties(libxl__gc *gc,
                                 const libxl_version_info *vers,
-                                void *fdt)
+                                void *fdt,
+                                const libxl_domain_build_info *info)
 {
-    int res;
+    int res, i;
+    size_t sz = 0;
+    char *compat, *p;
+    const char *compats[2 + info->num_compatibles];
+    const char *compat0 = GCSPRINTF("xen,xenvm-%d.%d", vers->xen_version_major, vers->xen_version_minor);
+    const char *compat1 = "xen,xenvm";
 
     res = fdt_property_string(fdt, "model", GCSPRINTF("XENVM-%d.%d",
                                                       vers->xen_version_major,
                                                       vers->xen_version_minor));
     if (res) return res;
 
-    res = fdt_property_compat(gc, fdt, 2,
-                              GCSPRINTF("xen,xenvm-%d.%d",
-                                        vers->xen_version_major,
-                                        vers->xen_version_minor),
-                              "xen,xenvm");
+    compats[0] = compat0;
+    compats[1] = compat1;
+    sz = strlen(compat0) + strlen(compat1) + 2;
+
+    for(i = 0; i < info->num_compatibles; i++)
+    {
+        compats[2 + i] =  info->compatible[i].compat;
+        sz += strlen(info->compatible[i].compat) + 1;
+    }
+
+    p = compat = libxl__zalloc(gc, sz);
+    for (i = 0; i < 2 + info->num_compatibles; i++) {
+        strcpy(p, compats[i]);
+        p += strlen(compats[i]) + 1;
+    }
+
+    res = fdt_property(fdt, "compatible", compat, sz);
     if (res) return res;
 
     res = fdt_property_cell(fdt, "interrupt-parent", PHANDLE_GIC);
@@ -847,7 +860,7 @@ next_resize:
 
         FDT( fdt_begin_node(fdt, "") );
 
-        FDT( make_root_properties(gc, vers, fdt) );
+        FDT( make_root_properties(gc, vers, fdt, info) );
         FDT( make_chosen_node(gc, fdt, !!dom->ramdisk_blob, state, info) );
         FDT( make_cpus_node(gc, fdt, info->max_vcpus, ainfo) );
         FDT( make_psci_node(gc, fdt) );
diff --git a/tools/libxl/libxl_types.idl b/tools/libxl/libxl_types.idl
index d8a4132..2e5611c 100644
--- a/tools/libxl/libxl_types.idl
+++ b/tools/libxl/libxl_types.idl
@@ -443,6 +443,10 @@ libxl_passthrough_info = Struct("passthrough_info", [
     ("path",       string),
     ])
 
+libxl_compat_info = Struct("compat_info", [
+    ("compat",       string),
+    ])
+
 libxl_domain_build_info = Struct("domain_build_info",[
     ("max_vcpus",       integer),
     ("avail_vcpus",     libxl_bitmap),
@@ -499,6 +503,7 @@ libxl_domain_build_info = Struct("domain_build_info",[
     # 65000 which is reserved by the toolstack.
     ("device_tree",      string),
     ("passthrough_nodes",            Array(libxl_passthrough_info, "num_passthrough_nodes")),
+    ("compatible",            Array(libxl_compat_info, "num_compatibles")),
     ("u", KeyedUnion(None, libxl_domain_type, "type",
                 [("hvm", Struct(None, [("firmware",         string),
                                        ("bios",             libxl_bios_type),
diff --git a/tools/libxl/xl_cmdimpl.c b/tools/libxl/xl_cmdimpl.c
index 35b5645..87c20e5 100644
--- a/tools/libxl/xl_cmdimpl.c
+++ b/tools/libxl/xl_cmdimpl.c
@@ -1325,7 +1325,7 @@ static void parse_config_data(const char *config_source,
     long l, vcpus = 0;
     XLU_Config *config;
     XLU_ConfigList *cpus, *vbds, *nics, *pcis, *cvfbs, *cpuids, *vtpms,
-                   *usbctrls, *usbdevs, *nodes;
+                   *usbctrls, *usbdevs, *nodes, *compats;
     XLU_ConfigList *channels, *ioports, *irqs, *iomem, *viridian, *dtdevs;
     int num_ioports, num_irqs, num_iomem, num_cpus, num_viridian;
     int pci_power_mgmt = 0;
@@ -1868,6 +1868,18 @@ static void parse_config_data(const char *config_source,
         }
     }
 
+    if (!xlu_cfg_get_list (config, "compatibles", &compats, 0, 0)) {
+        d_config->b_info.num_compatibles = 0;
+        d_config->b_info.compatible = NULL;
+        while ((buf = xlu_cfg_get_listitem (compats, d_config->b_info.num_compatibles)) != NULL) {
+            libxl_compat_info * compat;
+            compat = ARRAY_EXTEND_INIT_NODEVID(d_config->b_info.compatible,
+                                               d_config->b_info.num_compatibles,
+                                               libxl_compat_info_init);
+            compat->compat = strdup(buf);
+        }
+    }
+
     if (!xlu_cfg_get_list(config, "vtpm", &vtpms, 0, 0)) {
         d_config->num_vtpms = 0;
         d_config->vtpms = NULL;
-- 
1.9.1

